build/builder.go

```diff
@@ -222,6 +222,8 @@ type Builder struct {
 
 	// a sortable 20 chars long id.
 	id string
+
+	existingShards []string
```

Comment:
Sorry, you're correct (but this check is still necessary). The failure mode isn't the shard writing, it's that `gitIndex` calls `b.Finish()` twice in the same run. Once at the end of the function, and another time in a defer statement: 

https://github.com/sourcegraph/zoekt/blob/86e6a54/gitindex/index.go#L466-L468 

The first time b.finish() is invoked, all the shards that it discovers through `oldShards := b.FindAllShards()` must be shards that existed prior to this latest job (since it hasn't renamed any of the new shards yet). So, it's safe to update the `FileTombstones` information on all of `oldShards`. 

However, the second time b.Finish() runs (as a result of the defer), without `existingShards` it can't disambiguate between the shards that it spat out during the previous b.Finish() invocation and shards that existed prior to this latest indexing job. In this case, it'll ended up tombstoning files in the shards that it just created (which means that the latest version of those files are never discoverable). 

That's why we have to cache the list of previous existing shards in `existingShards` during builder's instantiation. This list is only written to once, and is never updated regardless of how many times you call b.Finish()`. Only those shards should get their filetombstone metadata updated.

Reply:
But Finish is supposed to be idempotent, which is why previously we just called it twice. Can you just make sure that Finish stays idempotent then you don't need this extra state.